src/common/framework/gpu/device_pool.ts

```diff
@@ -1,56 +1,61 @@
-import { assert, raceWithRejectOnTimeout, unreachable, assertReject } from '../util/util.js';
+import { SkipTestCase } from '../fixture.js';
+import {
+  assert,
+  raceWithRejectOnTimeout,
+  unreachable,
+  assertReject,
+  objectEquals,
+} from '../util/util.js';
 
 import { getGPU } from './implementation.js';
 
-interface DeviceHolder {
-  acquired: boolean; // whether the device is currently in use by a test
-  device: GPUDevice;
-  lostReason?: string; // initially undefined; becomes set when the device is lost
+export interface DeviceProvider {
+  acquire(): GPUDevice;
 }
 
 class TestFailedButDeviceReusable extends Error {}
 export class TestOOMedShouldAttemptGC extends Error {}
 
-const kPopErrorScopeTimeoutMS = 5000;
-
 export class DevicePool {
-  failed: boolean = false; // if device init failed once, never try again
-  holder?: DeviceHolder = undefined; // undefined if "uninitialized" (not yet initialized, or lost)
+  /** Device with no descriptor. */
+  private defaultHolder: DeviceHolder | 'uninitialized' | 'failed' = 'uninitialized';
```

Comment:
why `"failed" = "uninitialized"`?

Reply:
the type is `DeviceHolder | 'uninitialized' | 'failed'`, the initial value is `'uninitialized'`

Reply:
Gotcha